### PR TITLE
Add gcloud-based GCSHook

### PR DESCRIPTION
**Pull request by jlowin**
A GCSHook based on the `gcloud` library, analogous to the `S3Hook` in the main airflow package.

see conversation in #1109

The file itself lives in `contrib/gcloud/` to clearly identify it as different from the current GCSHook.

### CODE REVIEW

**Comment by jlowin** [comment]
follows #1118

**Comment by criccomini** [comment]
Could you make it so we can specify the service account from the JSON? We actually run Airflow with multiple service accounts per-hook to get fine-grained permissions. Setting it at the environment variable level doesn't work well with this.

File: setup.py [hunk 1]
```diff
@@ -53,6 +53,14 @@ def run(self):
 ]
 docker = ['docker-py>=1.6.0']
 druid = ['pydruid>=0.2.1']
+gcloud = [
+    'gcloud>=1.1.0',
+]
+gcp_api = [
+    'httplib2',
+    'google-api-python-client',
```
Comment (jlowin) [inline setup.py:61]: @criccomini I think that for compatibility this might be better as `google-api-python-client<1.5` but I'm not sure. As far as I can tell 1.5 is the first version that depends on `oauth2client 2.0`.

File: requirements.txt [hunk 2]
```diff
@@ -16,16 +16,13 @@ flask-cache
 flask-login
 flower
 future
-google-api-python-client
```
Comment (criccomini) [inline requirements.txt:19]: This stuff can go now. I merged #1118

**Comment by criccomini** [comment]
What do you think about migrating the GCS operators to use new hook as part of this?

**Comment by jlowin** [comment]
@criccomini regarding service accounts it should work -- this Hook looks for a "keyfile_path"  in the airflow connection to identify a json service account key. I think there's similar extra parameter for the other GCS Hook. However this one uses json and I think the other one uses p12 if I remember correctly. I am going to make a change so you can pass either a json OR a p12 file to this hook, so you could reuse the same GCS connections.

**Comment by criccomini** [comment]
Ah, got it. Missed that. Awesome. Motivation to switch to json. :) We're still on the deprecated p12's for the most part.

**Comment by jlowin** [comment]
No problem, will support both! :) One sec.

**Comment by jlowin** [comment]
@criccomini I think this will now allow you to reuse your existing Airflow connections, but please give it a shot to be sure.

**Comment by criccomini** [comment]
> @criccomini I think this will now allow you to reuse your existing Airflow connections, but please give it a shot to be sure.

Thought: perhaps add backwards-compatible APIs so that the existing GCS operators can work with your new hooks?

**Comment by jlowin** [comment]
I think that's the cleanest approach -- let me see how easily I can make the signatures match.

**Comment by jlowin** [comment]
@criccomini I'm feeling pretty good about this! the latest one _should_ have full signature compatibility with the previous hook, both at initialization and for the `download` and `upload` methods. Also full support for scopes, which we didn't have before (though note scopes require passing key_path either in the connection or at initialization).

**Comment by jlowin** [comment]
...except for "delegate_to". I'm not familiar with this, unfortunately. You can pass the argument, but it's ignored.

With regard to other operators, at this time our GCP use is focused on GCS/GCE/GKE/CloudSQL/DataProc so I don't think I have the expertise to migrate them. However we are starting to use BigQuery more so maybe that'll change as our workflow evolves.

**Comment by criccomini** [comment]
@jlowin is this ready for another round of reviews?

**Comment by jlowin** [comment]
@criccomini yes

File: airflow/www/static/connection_form.js [hunk 3]
```diff
@@ -7,7 +7,12 @@
         jdbc: {
             hidden_fields: ['port', 'schema', 'extra'],
             relabeling: {'host': 'Connection URL'},
+        },
+        google_cloud_platform: {
+            hidden_fields: ['host', 'schema', 'login', 'password', 'port', 'extra'],
```
Comment (criccomini) [inline airflow/www/static/connection_form.js:12]: Don't we still need 'extra'?

File: airflow/www/static/connection_form.js [hunk 4]
```diff
@@ -7,7 +7,12 @@
         jdbc: {
             hidden_fields: ['port', 'schema', 'extra'],
             relabeling: {'host': 'Connection URL'},
+        },
+        google_cloud_platform: {
```
Comment (criccomini) [inline airflow/www/static/connection_form.js:11]: Not sure if this will work with data profiler. Right now, we have different types for BQ, GCS, datastore, etc. This allows us to set data profiler to query BigQuery. This code is in models.py's get_hook method:

```
        elif self.conn_type == 'bigquery':
            return contrib_hooks.BigQueryHook(bigquery_conn_id=self.conn_id)
```

I suppose we could assume it's a BQ hook if the connection is of type 'google_cloud_platform'. But it's worth considering.

Comment (criccomini) [inline airflow/www/static/connection_form.js:11]: (note, this is a forward looking comment, for when we migrate BQ to the new gcloud based hook)

File: airflow/www/static/connection_form.js (hunk 3 above)
Comment (jlowin) [inline airflow/www/static/connection_form.js:12]: No, new connections should be set up using "project", "key_path", etc. fields that are displayed in the Airflow UI, not extras. I think "extras" is opaque for users who haven't read the code. However, if someone uses a connection that predates this change and DOES have an "extra" field, the code [here](https://github.com/airbnb/airflow/pull/1119/files#diff-a9d313c95097c265b6b934953bdfb2dbR78) will recognize that and handle it properly.

As far as I know, the connection type is largely irrelevant to Airflow -- you could create an "FTP" connection and pass it to `GCSHook`, and as long as it had the correct attributes it would work properly. The only thing that the type does affect is the fields that are displayed/handled by the UI (which are put in the "extras" field anyway behind the scenes)

So bottom line, lets say you have a connection of type `google_cloud_storage` with `key_path` in extras, and you've been using it with `GoogleCloudStorageHook`, you should be able to pass that same connection to a `GCSHook` no problem.

Comment (criccomini) [inline airflow/www/static/connection_form.js:12]: Cool. See note about connection type, above, though. Data profiler does use it.

**Comment by criccomini** [comment]
I'm going to merge this in. I think that it will be safe to assume that a GCP connection type should be treated as BQ in data profiler.
